server/src/main/java/org/elasticsearch/action/support/local/TransportLocalClusterStateAction.java

```diff
@@ -0,0 +1,129 @@
+        this.executor = executor;
+    }
+
+    protected abstract ClusterBlockException checkBlock(Request request, ClusterState state);
+
+    protected abstract void localClusterStateOperation(Task task, Request request, ClusterState state, ActionListener<Response> listener)
+        throws Exception;
+
+    @Override
+    protected final void doExecute(Task task, Request request, ActionListener<Response> listener) {
+        final var state = clusterService.state();
+        final var clusterBlockException = checkBlock(request, state);
+        if (clusterBlockException != null) {
+            if (clusterBlockException.retryable() == false) {
+                listener.onFailure(clusterBlockException);
+            } else {
+                waitForClusterUnblock(task, request, listener, state, clusterBlockException);
+            }
+            return;
+        }
+
+        innerDoExecute(task, request, listener, state);
+    }
+
+    private void innerDoExecute(Task task, Request request, ActionListener<Response> listener, ClusterState state) {
+        if (task instanceof CancellableTask cancellableTask && cancellableTask.notifyIfCancelled(listener)) {
+            return;
+        }
+        // Workaround for https://github.com/elastic/elasticsearch/issues/97916 - TODO remove this when we can
+        executor.execute(ActionRunnable.wrap(listener, l -> localClusterStateOperation(task, request, state, l)));
+    }
+
+    private void waitForClusterUnblock(
+        Task task,
+        Request request,
+        ActionListener<Response> listener,
+        ClusterState initialState,
+        ClusterBlockException exception
+    ) {
+        var observer = new ClusterStateObserver(
+            initialState,
+            clusterService,
+            request.clusterUpdateTimeout(),
+            logger,
+            clusterService.threadPool().getThreadContext()
+        );
+        observer.waitForNextChange(new ClusterStateObserver.Listener() {
+            @Override
+            public void onNewClusterState(ClusterState state) {
+                logger.trace("retrying with cluster state version [{}]", state.version());
+                innerDoExecute(task, request, listener, state);
+            }
+
+            @Override
+            public void onClusterServiceClose() {
+                listener.onFailure(new NodeClosedException(clusterService.localNode()));
+            }
+
+            @Override
+            public void onTimeout(TimeValue timeout) {
+                logger.debug(
+                    () -> format("timed out while waiting for cluster to unblock in [%s] (timeout [%s])", actionName, timeout),
+                    exception
+                );
+                listener.onFailure(new ElasticsearchTimeoutException("timed out while waiting for cluster to unblock", exception));
+            }
+        }, clusterState -> isTaskCancelled(task) || checkBlock(request, clusterState) == null);
```

Comment:
This means that if the task is cancelled while waiting for an unblocking cluster state then it'll wait for the next cluster state update before completing, which could be arbitrarily far in the future. Could we use `org.elasticsearch.tasks.CancellableTask#addListener` to react more promptly to cancellation instead?

Reply:
Good idea :+1:. It looks like we're not doing that in `TransportMasterNodeAction` either. Am I missing something or should we add it there too?

Reply:
Eh we probably should. But that's a good point, let's leave this with the same behaviour as in `TransportMasterNodeAction` and we can come back to this change later.

Reply:
I opened https://github.com/elastic/elasticsearch/issues/117971 for this.